src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TimeoutTests.cs

```diff
@@ -0,0 +1,840 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+using Microsoft.AspNetCore.Testing;
+using Microsoft.Extensions.Logging.Testing;
+using Microsoft.Net.Http.Headers;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests
+{
+    public class Http3TimeoutTests : Http3TestBase
+    {
+        [Fact]
+        public async Task HEADERS_IncompleteFrameReceivedWithinRequestHeadersTimeout_AbortsConnection()
+        {
+            var timerStartedTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+            _mockTimeoutControl.Setup(tc => tc.SetTimeout(It.IsAny<long>(), TimeoutReason.RequestHeaders))
+                .Callback((long ticks, TimeoutReason reason) =>
+                {
+                    _timeoutControl.SetTimeout(ticks, reason);
+                    timerStartedTcs.SetResult();
+                });
+            var mockSystemClock = _serviceContext.MockSystemClock;
+            var limits = _serviceContext.ServerOptions.Limits;
+
+            _timeoutControl.Initialize(mockSystemClock.UtcNow.Ticks);
+
+            var requestStream = await InitializeConnectionAndStreamsAsync(_noopApplication);
+
+            var controlStream = await GetInboundControlStream();
+            await controlStream.ExpectSettingsAsync();
+
+            await timerStartedTcs.Task;
+
+            await requestStream.SendHeadersPartialAsync();
+
+            AdvanceClock(limits.RequestHeadersTimeout + Heartbeat.Interval);
+
+            _mockTimeoutHandler.Verify(h => h.OnTimeout(It.IsAny<TimeoutReason>()), Times.Never);
+
+            AdvanceClock(TimeSpan.FromTicks(1));
+
+            _mockTimeoutHandler.Verify(h => h.OnTimeout(TimeoutReason.RequestHeaders), Times.Once);
+
+            await WaitForConnectionStopAsync(expectedLastStreamId: 0, ignoreNonGoAwayFrames: false);
+
+            _mockTimeoutHandler.VerifyNoOtherCalls();
+        }
+
+        /*
```

Comment:
I've left commented out tests here so I can track converting what should be applied to HTTP/3.